specification/theory_operations.adoc

```diff
@@ -1,2 +1,430 @@
+domain manager, e.g. the host VMM. As a consequence, it is the host security
+domain's responsibility to reserve the MSI addresses and then rely on the TSM to
+program the IOMMUs with those reserved addresses. This IOMMU registration
+process is driven by the untrusted domain manager for all IOMMUs that
+participate in TEE-IO and operates as described in the following steps:
+
+1. The TSM is loaded into a supervisor domain and provisioned with a CoVE-IO
+   manifest. It is recommended that the TSM is measured by the platform
+   root-of-trust for subsequent attestation.
+2. The host supervisor domain manager (e.g. the host VMM) enumerates all
+   platform IOMMUs as PCIe devices.
+3. The host IOMMU driver is loaded and probed.
+4. The host IOMMU driver allocates MSI vectors for the trusted domain
+   IOMMUs. Those vectors must point to the untrusted IOMMU driver as the TSM can
+   not handle external interrupts.
+5. The host supervisor domain manager registers the IOMMUs with the TSM by
+   calling the `sbi_covh_register_iommu()` COVH function. The TSM gets the
+   allocated MSI vectors and configures the trusted IOMMU `s_msi_cfg_table`
+   register accordingly.
+6. When a trusted IOMMU sends an MSI, the untrusted IOMMU driver handles it and
+   notifies the TSM about a pending trusted IOMMU MSI by calling the
+   `sbi_covh_notify_iommu_msi()` COVH function.
+7. The TSM verifies that there is a pending MSI by reading the IOMMU `s_ipsr`
+   register, and handles the interrupts as needed.
+
+The host IOMMU driver may be malicious and attempt to trick the TSM by either
+invoking the `sbi_covh_notify_iommu_msi()` COVH function while there are no
+pending MSI from the trusted IOMMU or not invoking it when there actually is one
+such pending interrupt. In the former case, the TSM can verify that there really
+is a pending MSI by checking the trusted IOMMU status registers. The latter case
+could cause a denial of service, which is not in scope of the confidential
+computing threat model. If the TSM must ensure that a given command is completed
+without having to rely on untrusted IOMMU driver MSI notifications, it can queue
+an `IOFENCE.C` command after the desired command and check for the `cqh`
+advancing past the `IOFENCE.C` command index.
+
+==== PCIe Root Port Registration
+
+When the data link between the TDI and the TVM must be secured, trusted I/O
+relies on the PCIe IDE protocol. IDE provides confidentiality and integrity
+protection for TLPs received and transmitted between the physical device the TDI
+belongs to and its assigned PCIe root port (RP). Both endpoints (The RP and the
+physical device) must be configured with the same encryption keys through the
+IDE key management protocols and have their IDE PCIe extended capability
+configured as well.
+
+As described in the threat model and requirement sections, the following trusted
+I/O rules apply:
+
+* As PCIe switches are outside of the TVM TCB, only selective IDE streams are
```

Comment:
Do we need to program the switches when using selective IDE streams?

Reply:
For example, IDE has register - "IDE Control Register (Offset 08h)"
BIT2 - **Flow-Through IDE Stream Enable** - For Switch Ports and Root Ports, Enables the Port for flow-through operation of TLPs associated with Selective IDE streams.